invoke-ai/InvokeAI
# feat: generic progress events

## Summary

Generalized progress events:
- New `invocation_progress` event replaces `invocation_denoise_progress` event. It is emitted with a `message` and some optional args: `percentage`, `image`, `image_size`.
- Invocation API updated with `signal_progress` method.
- Updated SD, MultiDiffusion and FLUX to use new API. Also fixed some off-by-one kinda issues.
- Added progress to spandrel nodes, so we get a progress bar as these nodes execute.
- Updated UI to consume the new events.

## Related Issues / Discussions

- Supersedes #6715 (it would have been messy to update that PR so I made a new one with mostly the same changes)
- Supersedes #6777 (similar but this is more flexible for any progress event)

## QA Instructions

### How to test

Text matrix:
- SD1.5, SDXL, FLUX
- Send to Gallery, Send to Canvas
- Canvas Tab, Upscaling Tab

Set the log level to trace in system settings, open the browser JS console and generate. You should see:
- Console messages for each denoise event, starting from 0% and ending at 100%.
- Progress images per usual. You'll notice that FLUX now displays its initial latents.
- The progress bar should fill up at the last step.
- For upscaling, the progress bar should tick along during the spandrel step, and you should get console logs.

### Discussion items

- Maybe `signal_progress` isn't the best name for the API method.
- Maybe `image_size` should be `image_display_size`.
- I have not updated the `_new_invoke` step callback handling. I'm not sure how to trigger this code path to test it, but the changes should be similar to the ones I made in b48db16ca67ce063f6ba9c8323a99fd68f47fe38 - just adjust the `step` and `total_steps` parameters.

## Merge Plan

Would be nice to get this in for v5 but it's not super important.

## Checklist
